src/contexts/ServerContext.js

```diff
@@ -428,10 +428,27 @@ export const ServerProvider = ({ children }) => {
     const loadServers = async () => {
         setIsLoadingZoneContext(true);
         let zone_report = await loadZoneReport();
+
         if (zone_report !== undefined) {
             let resc_types = new Set();
-            let catalog_service_provider = [zone_report.data.zones[0]['icat_server']];
-            let fullServersArray = catalog_service_provider.concat(zone_report.data.zones[0]['servers']);
+            let version = zone_report.data.zones[0].servers[0]?.version?.irods_version;
+            let diffFrom4_3_1
+            try {
+                // 4.3.1 is a new release of schema after 4.1.0; this code handles both versions
+                diffFrom4_3_1 = irodsVersionComparator(version, "4.3.1"); // if version > 4.3.1, returns positive integer, if equal to 4.3.1, returns 0, if less than 4.3.1, returns negative
+            } catch(e) {
+                diffFrom4_3_1 = -1; // on versions before 4.3.1, version will be undefined
```

Comment:
can we make sure version is defined first?    no matter what?    not sure if that's 'better' or more pythonic, but this feels unnecessarily loose.

Reply:
We can define version like this: `let version = zone_report.data.zones[0].servers[0]?.version?.irods_version || zone_report.data.zones[0].icat_server?.version?.irods_version;`. This would handle both 4.3.1 (and after), and before 4.3.1 versions of zonereport

Reply:
would that... always grab `servers[0]`, even for 'before 4.3.1'?   isn't `servers` required?

Reply:
4-2-stable - https://github.com/irods/irods/blob/4-2-stable/configuration_schemas/v3/zone_bundle.json#L20

4.3.0 - https://github.com/irods/irods/blob/4.3.0/schemas/configuration/v4/zone_bundle.json.in#L20

Reply:
Yep, this code would always grab `servers[0]`. If there's an instance where the irods_version could be different among servers (which may make the first server's version different from the other servers' versions), this may (?) get messy. Looking for `icat_server`'s key in the object doesn't run into this issue though, I think.

Yes, `servers` is required, but before 4.3.1, the servers array could be empty because icat_server is stored separately. Thus, we can't have just  `let version = zone_report.data.zones[0].servers[0]?.version?.irods_version` by itself.

<img width="892" alt="Screenshot 2023-07-05 at 1 34 52 PM" src="https://github.com/irods/irods_client_zone_management_tool/assets/62436772/aaaff9ad-4dd6-4ffe-a1a2-8053f835c64e">

Reply:
got it - the `||` case catches when there is only one server in the zone for pre-4.3.1.

And *if* there is more than one server, then `servers[0]` will be populated... and we'll get `version` from that...

which is why you asked if the servers in a zone could be different versions... 

so... yes, they could be different versions.

Now... would that matter for this logic?  That's a separate question... I think maybe it does?

Perhaps we need a new approach... sniff on `v3` vs `v4`, and then if `v4`, we have a split on `icat_server` or not... and then branch / populate the list of servers accordingly.

Happy to hop on a call.

Reply:
From our discussion, I just pushed a commit that checks if we are version 4.3.1 or later by checking for the `icat_server` key (only present in pre-4.3.1 iRODS versions)